lucene/analysis/icu/src/test/org/apache/lucene/analysis/icu/TestICUTransformCharFilter.java

```diff
@@ -0,0 +1,153 @@
+  }
+  
+  public void testRollbackBuffer() throws Exception {
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"),
+        "яяяяя", "âââââ"); // final NFC transform applied
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"), 0,
+        "яяяяя", "a\u0302a\u0302a\u0302a\u0302a\u0302"); // final NFC transform never applied
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"), 2,
+        "яяяяя", "ââa\u0302a\u0302a\u0302");
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"), 4,
+        "яяяяяяяяяя", "ââa\u0302a\u0302a\u0302a\u0302a\u0302a\u0302a\u0302a\u0302");
+    checkToken(Transliterator.getInstance("Cyrillic-Latin"), 8,
+        "яяяяяяяяяяяяяяяяяяяя", "ââââââa\u0302ââââa\u0302ââââa\u0302âââ");
+  }
+
+  public void testCustomFunctionality() throws Exception {
+    String rules = "a > b; b > c;"; // convert a's to b's and b's to c's
+    checkToken(Transliterator.createFromRules("test", rules, Transliterator.FORWARD), "abacadaba", "bcbcbdbcb");
+  }
+  
+  public void testCustomFunctionality2() throws Exception {
+    String rules = "c { a > b; a > d;"; // convert a's to b's and b's to c's
+    checkToken(Transliterator.createFromRules("test", rules, Transliterator.FORWARD), "caa", "cbd");
+  }
+  
+  public void testOptimizer() throws Exception {
+    String rules = "a > b; b > c;"; // convert a's to b's and b's to c's
+    Transliterator custom = Transliterator.createFromRules("test", rules, Transliterator.FORWARD);
+    assertTrue(custom.getFilter() == null);
+    new ICUTransformCharFilter(new StringReader(""), custom);
+    assertTrue(custom.getFilter().equals(new UnicodeSet("[ab]")));
+  }
+  
+  public void testOptimizer2() throws Exception {
+    checkToken(Transliterator.getInstance("Traditional-Simplified; CaseFold"), 
+        "ABCDE", "abcde");
+  }
+  
+  public void testOptimizerSurrogate() throws Exception {
+    String rules = "\\U00020087 > x;"; // convert CJK UNIFIED IDEOGRAPH-20087 to an x
+    Transliterator custom = Transliterator.createFromRules("test", rules, Transliterator.FORWARD);
+    assertTrue(custom.getFilter() == null);
+    new ICUTransformCharFilter(new StringReader(""), custom);
+    assertTrue(custom.getFilter().equals(new UnicodeSet("[\\U00020087]")));
+  }
+
+  private void checkToken(Transliterator transform, String input, String expected) throws IOException {
+    checkToken(transform, ICUTransformCharFilter.DEFAULT_MAX_ROLLBACK_BUFFER_CAPACITY, input, expected);
+  }
+
+  private void checkToken(Transliterator transform, int maxRollbackBufferCapacity, String input, String expected) throws IOException {
+    final KeywordTokenizer input1 = new KeywordTokenizer();
+    input1.setReader(new ICUTransformCharFilter(new StringReader(input), transform, maxRollbackBufferCapacity));
+    assertTokenStreamContents(input1, new String[] { expected });
+  }
+  
+  /** blast some random strings through the analyzer */
+  public void testRandomStrings() throws Exception {
```

Comment:
Have you "beasted" this test? ie run it a lot?

Reply:
Depends on the definition of "a lot" :-) ... I haven't set out specifically to run this many times, but it helped me uncover a number of bugs/edge cases, and has been running consistently clean through repeated testing in the course of development. Because it works off of random input, it already runs a lot of input (1000 iterations); I just tried increasing the "iteration" count to Integer.MAX_VALUE, and it ran fine for as long as I was patient enough to wait for it, fwiw.

Reply:
Running these tests with the flag `-Dtests.multiplier=n` or something causes the `testRandomStrings*` methods to run over `n` times more random input. Run with n=100, no problems ...